homeassistant/components/heos/media_player.py

```diff
@@ -1,152 +1,261 @@
+    SUPPORT_PAUSE, SUPPORT_PLAY, SUPPORT_PREVIOUS_TRACK, SUPPORT_SHUFFLE_SET,
+    SUPPORT_STOP, SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
 from homeassistant.const import STATE_IDLE, STATE_PAUSED, STATE_PLAYING
+from homeassistant.util.dt import utcnow
 
-from . import DOMAIN as HEOS_DOMAIN
+from .const import DOMAIN as HEOS_DOMAIN
 
-DEPENDENCIES = ["heos"]
+DEPENDENCIES = ['heos']
 
-SUPPORT_HEOS = (
-    SUPPORT_PLAY
-    | SUPPORT_STOP
-    | SUPPORT_PAUSE
-    | SUPPORT_PLAY_MEDIA
-    | SUPPORT_PREVIOUS_TRACK
-    | SUPPORT_NEXT_TRACK
-    | SUPPORT_VOLUME_MUTE
-    | SUPPORT_VOLUME_SET
-    | SUPPORT_VOLUME_STEP
-)
+BASE_SUPPORTED_FEATURES = SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | \
+                          SUPPORT_VOLUME_STEP | SUPPORT_CLEAR_PLAYLIST | \
+                          SUPPORT_SHUFFLE_SET
 
-PLAY_STATE_TO_STATE = {
-    "play": STATE_PLAYING,
-    "pause": STATE_PAUSED,
-    "stop": STATE_IDLE,
-}
 
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Platform uses config entry setup."""
+    pass
 
-async def async_setup_platform(hass, config, async_add_devices,
-                               discover_info=None):
-    """Set up the HEOS platform."""
-    controller = hass.data[HEOS_DOMAIN][DOMAIN]
-    players = controller.get_players()
-    devices = [HeosMediaPlayer(p) for p in players]
-    async_add_devices(devices, True)
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add binary sensors for a config entry."""
+    players = hass.data[HEOS_DOMAIN][DOMAIN]
+    devices = [HeosMediaPlayer(player) for player in players.values()]
+    async_add_entities(devices, True)
 
 
 class HeosMediaPlayer(MediaPlayerDevice):
     """The HEOS player."""
 
     def __init__(self, player):
         """Initialize."""
+        from pyheos import const
+        self._media_position_updated_at = None
         self._player = player
+        self._signals = []
+        self._supported_features = BASE_SUPPORTED_FEATURES
+        self._play_state_to_state = {
+            const.PLAY_STATE_PLAY: STATE_PLAYING,
+            const.PLAY_STATE_STOP: STATE_IDLE,
+            const.PLAY_STATE_PAUSE: STATE_PAUSED
+        }
+        self._control_to_support = {
+            const.CONTROL_PLAY: SUPPORT_PLAY,
+            const.CONTROL_PAUSE: SUPPORT_PAUSE,
+            const.CONTROL_STOP: SUPPORT_STOP,
+            const.CONTROL_PLAY_PREVIOUS: SUPPORT_PREVIOUS_TRACK,
+            const.CONTROL_PLAY_NEXT: SUPPORT_NEXT_TRACK
+        }
+
+    async def _controller_event(self, event):
+        """Handle controller event."""
+        from pyheos import const
+        if event == const.EVENT_PLAYERS_CHANGED:
+            await self.async_update_ha_state(True)
+
+    async def _heos_event(self, event):
+        """Handle connection event."""
+        await self.async_update_ha_state(True)
```

Comment:
I think that you might want to use async_schedule_update_ha_state, so that you don't block heos doing their things.

Reply:
It won't block.  This is called via a dispatcher as a background task already.